### ISSUE TITLE
[Feature Request] - Add `ConnectivityHierarchicalClustering`

### ISSUE THREAD

**Issue by jolespin** [jolespin/skclust#36]
**Goal:** Refactor `HierarchicalClustering` into a base class + two subclasses, separating the full-distance-matrix path (fastcluster) from the connectivity-constrained path (sklearn).

**File:** `skclust/hierarchical.py` (single file, all three classes)

---

**Step 1: Create `BaseHierarchicalClustering`**

Extract all shared logic from the current `HierarchicalClustering`:

*Parameters (base `__init__`):*
- `method`, `metric`, `min_cluster_size`, `deep_split`, `cut_method`, `cut_threshold`, `name`, `random_state`, `cluster_prefix`
- New class attribute: `_valid_cut_methods` — subclasses set this to declare which cut methods they support

*Shared methods (move to base):*
- `fit()` — template method calling abstract hooks `_prepare_input(X)` and `_perform_clustering()`, then shared steps: dendrogram generation, `_cut_tree()`, label wrapping, `_build_tree()`
- `transform()`, `fit_transform()`
- `_cut_tree()` — dispatcher; validates `cut_method in self._valid_cut_methods`
- `_cut_tree_height()`, `_cut_tree_maxclust()`, `_cut_tree_dynamic_tree()`
- `_renumber_labels()` (static)
- `_apply_cluster_prefix()`
- `_build_tree()`
- `_check_fitted()`, `_check_plotting_available()`
- All plotting: `plot()`, `_plot_categorical_track()`, `_plot_tracks()`, `_generate_cluster_colors()`
- `add_track()`
- `summary()`

*Abstract hooks (subclasses must implement):*
- `_prepare_input(X)` — validate input, set `self.sample_labels_`, `self._has_sample_ids`
- `_perform_clustering()` — produce `self.linkage_matrix_`

*Dendrogram generation moves out of `_perform_clustering` into base `fit()`:*
```python
# In base fit(), after _perform_clustering():
self.dendrogram_ = scipy_dendrogram(self.linkage_matrix_, labels=self.sample_labels_, no_plot=True)
self.leaves_ = self.dendrogram_["ivl"]
```

*Validation in base `__init__`:*
- `method`, `min_cluster_size`, `cluster_prefix` — same as current
- `cut_method` validation deferred to `fit()` where `_valid_cut_methods` is checked
- `deep_split` validation stays parameterized by `cut_method`

---

**Step 2: Refactor `HierarchicalClustering(BaseHierarchicalClustering)`**

*Additional parameters (subclass `__init__`):*
- `pam_stage`, `pam_respects_dendro`, `use_medoids`, `max_pam_dist`, `respect_small_clusters` — hybrid-only
- `distance_matrix_tol`, `store_data`

*Set:* `_valid_cut_methods = ['hybrid', 'tree', 'height', 'maxclust']`

*Implement hooks:*
- `_prepare_input(X)` — current `_validate_input()` + distance matrix detection + `_compute_distance_matrix()`. Sets `self.distance_matrix_`, `self.sample_labels_`, `self._has_sample_ids`
- `_perform_clustering()` — `fastcluster.linkage` on `self.distance_matrix_.condensed_form()`. Sets `self.linkage_matrix_` only (dendrogram moves to base)

*Keep on this subclass only:*
- `_cut_tree_hybrid()` — needs `self.distance_matrix_`
- `_is_distance_matrix()`
- `_compute_distance_matrix()`
- `to_igraph()` — needs `self.distance_matrix_`
- `data_` storage

*Override `_cut_tree()`:* call hybrid if applicable, else `super()._cut_tree()`

---

**Step 3: Create `ConnectivityHierarchicalClustering(BaseHierarchicalClustering)`**

*Additional parameters:*
- `connectivity` — sparse matrix, dense array, or callable (matches sklearn's API)

*Set:* `_valid_cut_methods = ['tree', 'height', 'maxclust']`

*Implement hooks:*
- `_prepare_input(X)` — validate `X` is a feature matrix (not a distance matrix). If `connectivity` is callable, call it on `X`. Set `self.sample_labels_`, `self._has_sample_ids`
- `_perform_clustering()` — fit `sklearn.cluster.AgglomerativeClustering(n_clusters=None, distance_threshold=0, compute_full_tree=True, connectivity=self.connectivity_, affinity=self.metric, linkage=self.method)`, then reconstruct linkage matrix from `children_` + `distances_`

*New private method:*
- `_linkage_from_sklearn(model, n_samples)` — reconstructs scipy-format linkage matrix from sklearn's `children_`, `distances_`, and leaf counts

*Validation in `__init__`:*
- Raise `ValueError` if `cut_method='hybrid'` (with clear message explaining why)
- Raise `ValueError` if `metric='precomputed'` (connectivity path expects features, not distances)

---

**Step 4: Update exports**

Add `BaseHierarchicalClustering` and `ConnectivityHierarchicalClustering` to `__init__.py`.

---

**What doesn't change:** All downstream behavior — tree cutting, plotting, tracks, dendrogram, `summary()` — works identically for both subclasses since it all operates on `self.linkage_matrix_` and `self.sample_labels_`.

**Risk to watch:** sklearn's `AgglomerativeClustering` may not support all `method` values (it uses `linkage` param with `ward`, `complete`, `average`, `single`). `centroid`, `median`, `weighted` are fastcluster-only. Validate this in `ConnectivityHierarchicalClustering.__init__`.

**Comment by jolespin** [comment]
https://github.com/jolespin/skclust/pull/38
